middle_end/flambda/lift_code.ml

```diff
@@ -17,102 +17,339 @@
   let acc =
-    match let_expr with
+    match (let_expr : Flambda.let_expr) with
     | { var = v1; defining_expr = Expr (Let let2); _ } ->
-        let acc, body2 = extract_let_expr acc let2 in
-        Immutable(v1, W.expr body2) :: acc
+        extract_let_expr acc v1 let2 ~in_closure
     | { var = v1; defining_expr = Expr (Let_mutable let_mut); _ } ->
-        let acc, body2 = extract_let_mutable acc let_mut in
-        Immutable(v1, W.expr body2) :: acc
+        extract_let_mutable acc v1 let_mut ~in_closure
+    | { var = v1; defining_expr = Expr (Region expr); _ }
+          when should_lift_regions ~in_closure ->
+        extract_region acc v1 expr ~in_closure
+    | { var = v1;
+        defining_expr = Expr (Apply ({ reg_close = Rc_close_at_apply; _ }
+                                     as apply)) }
+          when should_lift_regions ~in_closure ->
+        extract_tail_call acc v1 apply
+    | { var = v1;
+        defining_expr = Expr (Send ({ reg_close = Rc_close_at_apply; _ }
+                                    as send)) }
+          when should_lift_regions ~in_closure ->
+        extract_tail_send acc v1 send
     | { var = v; _ } ->
         Immutable(v, W.of_defining_expr_of_let let_expr) :: acc
   in
   let body = W.of_body_of_let let_expr in
-  extract acc body
+  extract acc dest body ~in_closure
 
-and extract_let_mutable acc (let_mut : Flambda.let_mutable) =
+and extract_let_mutable acc dest let_mut ~in_closure =
   let module W = Flambda.With_free_variables in
   let { Flambda.var; initial_value; contents_kind; body } = let_mut in
   let acc = Mutable(var, initial_value, contents_kind) :: acc in
-  extract acc (W.of_expr body)
+  extract acc dest (W.of_expr body) ~in_closure
+
+(* C[let x = region M in x] *)
+and extract_region acc dest body ~in_closure =
+  let module W = Flambda.With_free_variables in
+  (* Ideally, we would recurse with the same [dest] since we're ultimately going
+     to store there anyway. Unfortunately, if [body] has an unliftable tail,
+     we're going to need [inner_dest] as an intermediary. *)
+  let inner_dest = Variable.rename dest in
+  (* C[let x = region (let y = M in y) in x] =>
+     C[let x = region (D[y]) in x]  *)
+  let acc_expr = extract [] inner_dest (W.of_expr body) ~in_closure in
+  (* If possible, recover the expression that gets assigned to [inner_dest] so
+     we can directly assign [dest] to it instead *)
+  match split_defs acc_expr inner_dest with
+  | acc_expr, body when liftable_region_body body ->
+    (* The accumulator must remain balanced between [Region] and [Exclave] (see
+       [check_defs]), since it defines a scope into which [extract_let_expr]
+       will move arbitrary computations - if there is a [Region] but no
+       [Exclave], this means we're moving those computations into a different
+       region. It may be that [acc_expr] already has an [Exclave] (because we
+       lifted it out of [body]), but otherwise we need to add it. *)
```

Comment:
Is a single `Exclave` definitely enough to fix this? Could there be two `Region`s?

Reply:
There's only one `Region` we need to worry about, and that's the one we're adding (on line 242). The extreme thing that `acc_expr` could do is _close_ more regions (that is, have more `Exclave`s), which will be handled by outer calls to `extract_region`.